onnxruntime/core/optimizer/shape_to_initializer.cc

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/shape_to_initializer.h"
+#include "core/graph/graph.h"
+#include "core/graph/graph_utils.h"
+#include "core/graph/op.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/optimizer_execution_frame.h"
+#include "core/framework/op_kernel.h"
+
+namespace onnxruntime {
+
+Status ShapeToInitializer::Apply(Graph& graph, Node& node, RewriteRuleEffect& rule_effect) const {
+  // Create an execution frame to get access to the memory allocator.
+  OptimizerExecutionFrame::Info info({&node}, InitializedTensorSet());
+
+  // Store the statically inferred shape of the input to the Shape operator.
+  const ONNX_NAMESPACE::TensorShapeProto* input_shape_proto = node.InputDefs()[0]->Shape();
+  std::vector<int64_t> input_dims;
+  int num_dimensions = input_shape_proto->dim_size();
+  for (int i = 0; i < num_dimensions; i++) {
+    ORT_ENFORCE(input_shape_proto->dim(i).has_dim_value());
+    input_dims.push_back(gsl::narrow_cast<int64_t>(input_shape_proto->dim(i).dim_value()));
+  }
+
+  // Create a tensor that corresponds to the output of the Shape operator.
+  const auto shape_tensor = std::make_unique<Tensor>(DataTypeImpl::GetType<int64_t>(),
+                                                     TensorShape({gsl::narrow_cast<int64_t>(num_dimensions)}),
+                                                     input_dims.data(), info.GetAllocator()->Info());
+
+  // Create the TensorProto that will be used as initializer in place of the Shape operator.
+  ONNX_NAMESPACE::TensorProto shape_initializer_proto;
+  const auto* shape_out_def = node.OutputDefs()[0];
+  graph_utils::BuildTensorProtoForInitializer(*shape_tensor, *shape_out_def, shape_initializer_proto);
+
```

Comment:
If you init shape_initializer_proto by copying shape_out_def, couldn't you just add the bytes for the dim values directly to raw data instead of doing things via Tensor?

Reply:
Done, I create the TensorProto directly.